src/nks/hcvault/provider.rs

```diff
@@ -0,0 +1,512 @@
+///     .await
+///     .expect("Failed to send key pair generation request");
+/// ```
+///
+/// # Errors
+///
+/// This function will return an error if:
+/// - The HTTP client cannot be built.
+/// - The API URL cannot be joined with the endpoint.
+/// - The request cannot be sent successfully.
+///
+/// The function uses `reqwest` for HTTP communication and handles any errors that may arise during the process.
+async fn get_and_save_key_pair_request(
+    token: &str,
+    key_name: &str,
+    key_type: &str,
+    nks_address: Url,
+    length: Option<u32>,
+) -> Result<reqwest::Response, Box<dyn std::error::Error>> {
+    let client = reqwest::Client::builder()
+        .danger_accept_invalid_certs(true)
+        .build()?;
+    let request_body = match length {
+        Some(len) => json!({
+            "token": token,
+            "name": key_name,
+            "type": key_type,
+            "length": len
+        }),
+        None => json!({
+            "token": token,
+            "name": key_name,
+            "type": key_type
+        }),
+    };
+    let api_url = nks_address.join("generateAndSaveKeyPair");
+    let response = client
+        .post(api_url.unwrap())
+        .header("accept", "*/*")
+        .header("Content-Type", "application/json-patch+json")
+        .json(&request_body)
+        .send()
+        .await?;
+    Ok(response)
+}
+
+/// Retrieves the secrets from the NksProvider.
+///
+/// This asynchronous function sends a POST request to the NksProvider's `getSecrets` endpoint.
+/// It then parses the JSON response and retrieves the `data` field which contains the secrets and the `newToken` field which contains the updated token.
+///
+/// # Arguments
+///
+/// * `token` - A string slice that represents the user token.
+/// * `nks_address_str` - A string slice that specifies the address of the NksProvider.
+///
+/// # Returns
+///
+/// A `Result` that, on success, contains an `Ok((String, String))`, where the first string is the JSON representation of the secrets and the second string is the updated token.
+/// On failure, it returns an `Err` with a `Box<dyn std::error::Error>`.
+///
+/// # Example
+///
+/// ```
+/// let nks_address_str = "https://nks.example.com";
+/// let runtime = Runtime::new().unwrap();
+/// match runtime.block_on(get_secrets("user_token", nks_address_str)) {
+///     Ok((secrets, new_token)) => {
+///         println!("Secrets: {}", secrets);
+///         println!("New token: {}", new_token);
+///     },
+///     Err(err) => println!("Failed to get secrets: {}", err),
+/// }
+/// ```
+async fn get_secrets(token: &str, nks_address_str: &str) -> anyhow::Result<(String, String), Box<dyn std::error::Error>> {
+    let client = reqwest::Client::builder()
+        .danger_accept_invalid_certs(true)
+        .build()?;
+    let body = json!({
+        "token": token
+    });
+
+    let response: Value = client.post(format!("{}getSecrets", nks_address_str))
```

Comment:
How do you authenticate against your ASP service?

Reply:
Our ASP API itself doesn’t handle the authentication; it relays the user token to the vault server, if the token is valid, then all the other operations follow, like saving keys, etc.